fix(ivy): move HostListeners generation to factory function

AndrewKushnir opened the pull request:
## PR Description
The goal of this PR is to move the generation of listeners defined as @HostListener from `hostBindings` to Factory function. See `factoryDef` in `ngtsc_spec.ts` (in this PR) for examples of how Factory function will look like in case of host listeners defined.

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

AndrewKushnir committed:
fix(ivy): move HostListeners generation to factory function

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview d038dae at https://pr26480-d038dae.ngbuilds.io/.

kara commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -483,8 +483,54 @@ describe('ngtsc behavioral tests', () => {
+      factory: function FooCmp_Factory(t) {
+        var f = new (t || FooCmp)();
+        i0.ɵlistener("click", function FooCmp_click_HostBindingHandler($event) {
+            var pd_b = (f.onClick($event) !== false);
```
It looks like you're generating old-style listeners (we don't need that false check). Check out the `compiler_listeners_spec` here for an example of what a listener should look like: https://github.com/angular/angular/blob/master/packages/compiler-cli/test/compliance/r3_view_compiler_listener_spec.ts#L96

We should probably also have a test in that spec.

kara commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -483,8 +483,54 @@ describe('ngtsc behavioral tests', () => {
+    expect(jsContents).toContain(factoryDef.replace(/\s+/g, ' ').trim());
+  });
+
+  it('should generate host bindings for directives with base factories', () => {
```
Sounds like you mean host listeners here?

kara commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -317,3 +335,24 @@ function isDelegatedMetadata(meta: R3FactoryMetadata): meta is R3DelegatedFactor
+    const handler = o.fn(
+        [new o.FnParam('$event', o.DYNAMIC_TYPE)],
+        [...bindingExpr.stmts, new o.ReturnStatement(bindingExpr.allowDefault)], o.INFERRED_TYPE,
+        null, functionName);
```
This needs to be updated to remove the extra `preventDefault` logic. See https://github.com/angular/angular/blob/master/packages/compiler/src/render3/view/template.ts#L973

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview 91ec7fe at https://pr26480-91ec7fe.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

AndrewKushnir commented:
Thanks for the review @kara! I've applied the necessary changes, updated tests and also added a runtime test to check listeners behavior with specified args. Please have a look when you get a chance. Thank you.

mary-poppins commented:
You can preview b1c25eb at https://pr26480-b1c25eb.ngbuilds.io/.

trotyl commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -483,8 +483,51 @@ describe('ngtsc behavioral tests', () => {
+          selector: '[test]',
+        })
+        class Dir extends Base {
+          @HostListener('onChange', ['arg'])
```
Would it better to be an event name following style guide (no-prefix-on)?

trotyl commented on packages/core/test/render3/listeners_spec.ts:
```diff
@@ -479,6 +479,42 @@ describe('event listeners', () => {
+          if (rf & RenderFlags.Create) {
+            elementStart(0, 'button');
+            {
+              listener('click', function() { return ctx.onClick(ctx.data.a, ctx.data.b); });
```
Unmatched parameters with template in comments above: `onClick(data.a, data.b, data.c)`

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

AndrewKushnir commented:
Thanks for the comments @trotyl. I've made the necessary changes.

mary-poppins commented:
You can preview 0fca4a8 at https://pr26480-0fca4a8.ngbuilds.io/.

kara commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -44,11 +44,13 @@ function baseDirectiveFields(
     type: meta.type,
     deps: meta.deps,
-    injectFn: R3.directiveInject,
+    injectFn: R3.directiveInject, hostListeners
```
NIt: Fix formatting (`hostListeners` on its own line)

kara commented on packages/core/test/render3/listeners_spec.ts:
```diff
@@ -479,6 +479,42 @@ describe('event listeners', () => {
+      });
+    }
+
+    const comp = renderComponent(MyComp);
```
Can you use `ComponentFixture` here? The `renderComponent` test util is deprecated, so we don't want to add any new tests with it.

AndrewKushnir replied:
I've also updated other places in this test suite to transition to `ComponentFixture`.

kara reviewed (approved):
Looks good, just two small things

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview f9b04df at https://pr26480-f9b04df.ngbuilds.io/.

AndrewKushnir commented:
@alxhub could you please have a look at this PR too? Need your approval in order to merge it.

alxhub commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -49,6 +52,11 @@ export interface R3ConstructorFactoryMetadata {
+  /**
+   * List of host listeners defined for a given type.
+   */
+  hostListeners?: ParsedEvent[]|null;
```
I don't think the factory generation code should know about host listeners or how to generate them. Instead, it should have a general mechanism for inserting statements into factory functions.

For example, that could be:

```
extraStatementFn: ((instance: o.Expression) => o.Statement[])|null;
```

Then, a consumer that wishes to generate extra statements would pass a callback that gets called with an `o.Expression` that refers to the newly constructed instance, and can return whatever statements they like to be added to the factory function.

Additionally, it should not be possible to declare a `R3DelegatedFnOrClassMetadata` or `R3ExpressionFactoryMetadata` value with the `extraStatementFn` if it's not supported. This may mean extracting out a base interface and making the others not inherit from `R3ConstructorFactoryMetadata`.

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview 9827f43 at https://pr26480-9827f43.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

AndrewKushnir commented:
@alxhub thanks for the review! I've updated the code based on your proposal. Please have a look when you get a chance. Thank you.

mary-poppins commented:
You can preview 4c532ba at https://pr26480-4c532ba.ngbuilds.io/.

alxhub commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -49,6 +49,11 @@ export interface R3ConstructorFactoryMetadata {
+  /**
+   * Function that allows extra statements to be inserted into factory function.
+   */
+  extraStatementFn?: ((instance: o.Expression) => o.Statement[])|null;
```
This parameter shouldn't be optional, it's already nullable.

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview bf0a2cb at https://pr26480-bf0a2cb.ngbuilds.io/.

AndrewKushnir committed:
fixup! fix(ivy): move HostListeners generation to factory function

mary-poppins commented:
You can preview db788e8 at https://pr26480-db788e8.ngbuilds.io/.

mary-poppins commented:
You can preview 53b0c84 at https://pr26480-53b0c84.ngbuilds.io/.

AndrewKushnir commented:
It looks like the `test_docs_examples_0` test is red on master, so this PR needs merge assistance.
